chain-storage/src/lib.rs

```diff
@@ -1,1079 +1,809 @@
-            codec.put_u64(self.id.0)?;
-            codec.put_u64(self.parent.0)?;
-            codec.put_u32(self.date.0)?;
-            codec.put_u32(self.date.1)?;
-            codec.put_u64(self.chain_length.0)?;
-            codec.put_u64(self.data.len() as u64)?;
-            codec.put_bytes(&self.data)?;
-            Ok(())
-        }
+    if let Some(old_block_id) = maybe_old_block_id {
+        let info_bin = info.get(old_block_id.clone())?.unwrap();
+        let mut block_info = BlockInfo::deserialize(&info_bin[..], id_size, old_block_id.to_vec());
+        block_info.remove_ref();
+        let info_bin = block_info.serialize();
+        info.insert(block_info.id().as_ref(), info_bin)?;
     }
 
-    impl chain_core::property::Deserialize for Block {
-        type Error = std::io::Error;
-
-        fn deserialize<R: std::io::BufRead>(reader: R) -> Result<Self, Self::Error> {
-            let mut codec = Codec::new(reader);
-            Ok(Self {
-                id: BlockId(codec.get_u64()?),
-                parent: BlockId(codec.get_u64()?),
-                date: BlockDate(codec.get_u32()?, codec.get_u32()?),
-                chain_length: ChainLength(codec.get_u64()?),
-                data: {
-                    let length = codec.get_u64()?;
-                    codec.get_bytes(length as usize)?.into_boxed_slice()
-                },
-            })
-        }
-    }
+    Ok(Ok(()))
 }
 
-#[cfg(test)]
-pub mod tests {
-    use super::test_utils::{Block, BlockId};
-    use super::*;
-    use chain_core::property::{Block as _, BlockId as _};
-    use rand_core::{OsRng, RngCore};
-
-    const SIMULTANEOUS_READ_WRITE_ITERS: usize = 50;
-
-    pub fn pick_from_vector<'a, A, R: RngCore>(rng: &mut R, v: &'a [A]) -> &'a A {
-        let s = rng.next_u32() as usize;
-        // this doesn't need to be uniform
-        &v[s % v.len()]
-    }
-
-    pub fn generate_chain<R: RngCore>(
-        rng: &mut R,
-        store: &mut BlockStoreConnection<Block>,
-    ) -> Vec<Block> {
-        let mut blocks = vec![];
-
-        let genesis_block = Block::genesis(None);
-        store.put_block(&genesis_block).unwrap();
-        blocks.push(genesis_block);
-
-        for _ in 0..10 {
-            let mut parent_block = pick_from_vector(rng, &blocks).clone();
-            let r = 1 + (rng.next_u32() % 9999);
-            for _ in 0..r {
-                let block = parent_block.make_child(None);
-                store.put_block(&block).unwrap();
-                parent_block = block.clone();
-                blocks.push(block);
+#[inline]
+#[allow(clippy::too_many_arguments)]
+fn remove_tip_impl(
+    blocks: &TransactionalTree,
+    info: &TransactionalTree,
+    chain_length_to_block_ids: &TransactionalTree,
+    tips: &TransactionalTree,
+    permanent_store: &PermanentStore,
+    block_id: &[u8],
+    root_id: &[u8],
+    id_size: usize,
+) -> Result<Result<RemoveTipResult, Error>, ConflictableTransactionError<()>> {
```

Comment:
Maybe, for everyone's sanity, make it a one-level `Result`, with an internal enum for the `Ok` value to carry outcomes considered non-fatal for the transaction by the logic of the `remove_tip` method.

Reply:
Apparently, this can be rewritten to `Result<RemoveTipResult, ConflictableTransactionError<Error>>` in the latest version of `sled` and I will do so.